### PR TITLE
Derived Gauge: Add support for extracting value from a function

### LINKED ISSUE #463: Derived Gauge: Add support for function based extractor
The `createTimeSeries ` method should allow users to pass (anonymous) function, that it calls to calculate or extract the gauge value whenever metrics are collected.

### PR DESCRIPTION
**Pull request by mayurkale22**
Fixes #463

### CODE REVIEW

File: examples/gauge/derived-gauge-quickstart.js [hunk 1]
```diff
@@ -54,6 +54,8 @@ const gauge = metricRegistry.addDerivedInt64Gauge('active_handles_total', metric
 const queue = new QueueManager();
 queue.addJob();
 
-// The value of the gauge is observed from the obj whenever metrics are
+// The value of the gauge is observed from a function whenever metrics are
 // collected. In this case it will be 1.
-gauge.createTimeSeries(labelValues, queue);
+gauge.createTimeSeries(labelValues, () => {
```
Comment (draffensperger) [inline examples/gauge/derived-gauge-quickstart.js]: Can this just be `queue.getPendingJobs` with no need for a wrapper arrow function?

File: packages/opencensus-core/src/metrics/gauges/derived-gauge.ts [hunk 2]
```diff
@@ -159,16 +163,18 @@ export class DerivedGauge implements types.Meter {
...
-    } else if (DerivedGauge.isLengthAttributeInterface(obj)) {
-      this.extractor = () => obj.length;
-    } else if (DerivedGauge.isLengthMethodInterface(obj)) {
-      this.extractor = () => obj.length();
-    } else if (DerivedGauge.isSizeAttributeInterface(obj)) {
-      this.extractor = () => obj.size;
-    } else if (DerivedGauge.isSizeMethodInterface(obj)) {
-      this.extractor = () => obj.size();
+    if (typeof objOrFn === 'function') {
+      this.extractor = () => objOrFn();
```
Comment (draffensperger) [inline packages/opencensus-core/src/metrics/gauges/derived-gauge.ts]: Can you just say `this.extractor = objOrFun`?

Comment (mayurkale22) [reply packages/opencensus-core/src/metrics/gauges/derived-gauge.ts]: To allow this I have to add new type (`ValueExtractorOrFunction`) which is type of `Function|ValueExtractor`. Let me know if this is what is desired.

Comment (draffensperger) [reply packages/opencensus-core/src/metrics/gauges/derived-gauge.ts]: Why is that? I would have assumed that the internal `extractor` variable would always of type `() => number` (a function that returns a number). Then these different if statements are mapping from the various shapes of input to a simple function that returns a number. So in the case when the input is a function, could we just assign it directly?

Or am I missing something here?

File: examples/gauge/derived-gauge-quickstart.js (hunk 1 above)
Comment (mayurkale22) [reply examples/gauge/derived-gauge-quickstart.js]: Done.

File: packages/opencensus-core/src/metrics/gauges/derived-gauge.ts (hunk 2 above)
Comment (mayurkale22) [reply packages/opencensus-core/src/metrics/gauges/derived-gauge.ts]: Defining a strongly-typed function solved this problem.

Comment (draffensperger) [reply packages/opencensus-core/src/metrics/gauges/derived-gauge.ts]: Cool!

**Pull request merged**